## <h1 align="center"> The Battle of Neighborhood | London Destinations Chooser <h1/>

### 1. Introduction

The Romans built the city of Londinium along the River Thames in the year AD 43 The name Londinium (and later 'London') came from the Celtic language of the Ancient Britons. In the year AD 61, the city was attacked and destroyed. Then the Romans rebuilt the city, and London became an important trading hub. After the decline of the Roman Empire, few people remained in London. This was partly because the Anglo-Saxon people of Sub-Roman Britain were primarily agricultural. Once the Romans had gone, trade with Continental Europe dwindled. In the 9th century, more people started living in London again. It became the largest city in England. However, it did not become the capital city of England until the 12th century. 

After the railways were built, London grew very big. Greater London has 33 London Boroughs (Neighborhoods) and a mayor. 
The old City of London is only a square mile in size but has its own Lord Mayor.

Another famous old part of Greater London is Westminster, which was always a different city from the City of London. In Westminster is Westminster Abbey (a cathedral), The Palace of Westminster (the Houses of Parliament, with Big Ben), and 10 Downing Street (where the Prime Minister lives).

A lot has changed over the years and we now take a look at how London city have grown.

London is quite the popular tourist and vacation destinations for people all around the world. It is diverse and multicultural and offer a wide variety of experiences that is widely sought after. We try to group the neighborhoods of London respectively and draw insights to what they look like now.


### 2. Business Problem

The aim is to help tourists choose their destinations depending on the experiences that the neighborhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to London or even if they want to relocate neighborhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer.

### 3. Data Description

We require geographical location data for London. Postal codes in each city serve as a starting point. Using Postal codes we use can find out the neighborhoods, boroughs, venues and their most popular venue categories. 

#### 3.1. Wikipedia

To derive our solution, we scrape our data from 

https://en.wikipedia.org/wiki/List_of_areas_of_London

This Wikipedia page has information about all the neighborhoods, we limit it London.

London borough: Name of Neighborhood

Post town: Name of borough

Postcode district: Postal codes for London.

This Wikipedia page lacks information about the geographical locations. To solve this problem we use ArcGIS API.

#### 3.2. ArcGIS API

ArcGIS Online enables you to connect people, locations, and data using interactive maps. Work with smart, data-driven styles and intuitive analysis tools that deliver location intelligence. Share your insights with the world or specific groups.

More specifically, we use ArcGIS to get the geo locations of the neighborhoods of London. The following columns are added to our initial dataset which prepares our data.

latitude: Latitude for Neighborhood

longitude: Longitude for Neighborhood

#### 3.3. Foursquare API Data

We will need data about different venues in different neighborhoods of that specific borough. In order to gain that information we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API.

After finding the list of neighborhoods, we then connect to the Foursquare API to gather information about venues inside each and every neighborhood. For each neighborhood, we have chosen the radius to be 500 meters.

The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

Neighborhood: Name of the Neighborhood

Neighborhood Latitude: Latitude of the Neighborhood

Neighborhood Longitude: Longitude of the Neighborhood

Venue: Name of the Venue

Venue Latitude: Latitude of Venue

Venue Longitude: Longitude of Venue

Venue Category: Category of Venue

#### 4. Methodology

#### Installing and Importing Python Libraries and Dependencies

In [3]:
!pip install geocoder
!pip install folium

     |████████████████████████████████| 98 kB 6.9 MB/s  eta 0:00:01
     |████████████████████████████████| 93 kB 2.6 MB/s  eta 0:00:01


Importing Libraries

In [5]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print("All Required Libraries Imported!")

All Required Libraries Imported!


The approach taken here is to explore each of the cities individually, plot the map to show the neighbourhoods being considered and then build our model by clustering all of the similar neighbourhoods together and finally plot the new map with the clustered neighbourhoods. We draw insights and then compare and discuss our findings.

### Exploring London

#### Neighbourhoods of London

We begin to start collecting and refining the data needed for the our business solution to work.

#### Data Collection

To get the neighbourhoods in london, we start by scraping the list of areas of london wiki page.

In [6]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

Response 200 means that we are able to make the connection

In [7]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
                                            Location  \
 0                                        Abbey Wood   
 1                                             Acton   
 2                                         Addington   
 3                                        Addiscombe   
 4                                       Albany Park   
 5                                  Aldborough Hatch   
 6                                           Aldgate   
 7                                           Aldwych   
 8                                          Alperton   
 9                                           Anerley   
 10                                            Angel   
 11                                        Aperfield   
 12                                          Archway   
 13                                   Ardleigh Green 

Scraping the webpage gives us all the tables present on the page. We need the 2nd table, so selecting the 2nd table.

In [8]:
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location  \
0                                        Abbey Wood   
1                                             Acton   
2                                         Addington   
3                                        Addiscombe   
4                                       Albany Park   
5                                  Aldborough Hatch   
6                                           Aldgate   
7                                           Aldwych   
8                                          Alperton   
9                                           Anerley   
10                                            Angel   
11                                        Aperfield   
12                                          Archway   
13                                   Ardleigh Green   
14                                           Arkley   
15                                      Arnos Grove   
16                                           Balham   
17                                         Bankside   
18                                         Barbican   
19                                          Barking   
20                                      Barkingside   
21                                       Barnehurst   
22                                           Barnes   
23                                      Barnes Cray   
24                                      Barnet Gate   
25       Barnet (also Chipping Barnet, High Barnet)   
26                                        Barnsbury   
27                                        Battersea   
28                                        Bayswater   
29                                        Beckenham   
30                                          Beckton   
31                                        Becontree   
32                                  Becontree Heath   
33                                       Beddington   
34                                     Bedford Park   
35                                        Belgravia   
36                                       Bellingham   
37                                          Belmont   
38                                          Belmont   
39                                     Belsize Park   
40                                        Belvedere   
41                                       Bermondsey   
42                                       Berrylands   
43                                    Bethnal Green   
44         Bexley (also Old Bexley, Bexley Village)   
45               Bexleyheath (also Bexley New Town)   
46                                          Bickley   
47                                      Biggin Hill   
48                                         Blackfen   
49                                      Blackfriars   
50                                       Blackheath   
51                        Blackheath Royal Standard   
52                                        Blackwall   
53                                          Blendon   
54                                       Bloomsbury   
55                                       Botany Bay   
56                                     Bounds Green   
57                                              Bow   
58                                       Bowes Park   
59                                        Brentford   
60                                      Brent Cross   
61                                       Brent Park   
62                                        Brimsdown   
63                                          Brixton   
64                                         Brockley   
65                                          Bromley   
66                    Bromley (also Bromley-by-Bow)   
67                                   Bromley Common   
68                                         Brompton   
69                                      Brondesbury   
70                                   Brunswick Park   
71                                      Bulls Cross   
72                          

### Data Preprocessing

we remove the spaces in the column titles and then we add _ between words.

In [10]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location  \
0                                        Abbey Wood   
1                                             Acton   
2                                         Addington   
3                                        Addiscombe   
4                                       Albany Park   
5                                  Aldborough Hatch   
6                                           Aldgate   
7                                           Aldwych   
8                                          Alperton   
9                                           Anerley   
10                                            Angel   
11                                        Aperfield   
12                                          Archway   
13                                   Ardleigh Green   
14                                           Arkley   
15                                      Arnos Grove   
16                                           Balham   
17                                         Bankside   
18                                         Barbican   
19                                          Barking   
20                                      Barkingside   
21                                       Barnehurst   
22                                           Barnes   
23                                      Barnes Cray   
24                                      Barnet Gate   
25       Barnet (also Chipping Barnet, High Barnet)   
26                                        Barnsbury   
27                                        Battersea   
28                                        Bayswater   
29                                        Beckenham   
30                                          Beckton   
31                                        Becontree   
32                                  Becontree Heath   
33                                       Beddington   
34                                     Bedford Park   
35                                        Belgravia   
36                                       Bellingham   
37                                          Belmont   
38                                          Belmont   
39                                     Belsize Park   
40                                        Belvedere   
41                                       Bermondsey   
42                                       Berrylands   
43                                    Bethnal Green   
44         Bexley (also Old Bexley, Bexley Village)   
45               Bexleyheath (also Bexley New Town)   
46                                          Bickley   
47                                      Biggin Hill   
48                                         Blackfen   
49                                      Blackfriars   
50                                       Blackheath   
51                        Blackheath Royal Standard   
52                                        Blackwall   
53                                          Blendon   
54                                       Bloomsbury   
55                                       Botany Bay   
56                                     Bounds Green   
57                                              Bow   
58                                       Bowes Park   
59                                        Brentford   
60                                      Brent Cross   
61                                       Brent Park   
62                                        Brimsdown   
63                                          Brixton   
64                                         Brockley   
65                                          Bromley   
66                    Bromley (also Bromley-by-Bow)   
67                                   Bromley Common   
68                                         Brompton   
69                                      Brondesbury   
70                                   Brunswick Park   
71                                      Bulls Cross   
72                          

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

#### Feature Selection

We need only the boroughs, Postal codes, Post town for further steps. We can drop the locations, dial codes and OS grid.

In [12]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

In [13]:
df1.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

let's rename the Postcode district column and the london borough to something simpler

In [14]:
df1.columns = ['Neighbourhood','town','post_code']
df1

Neighbourhood  \
0                               Bexley, Greenwich [7]   
1                   Ealing, Hammersmith and Fulham[8]   
2                                          Croydon[8]   
3                                          Croydon[8]   
4                                              Bexley   
5                                        Redbridge[9]   
6                                            City[10]   
7                                     Westminster[10]   
8                                           Brent[11]   
9                                         Bromley[11]   
10                                       Islington[8]   
11                                        Bromley[11]   
12                                      Islington[12]   
13                                       Havering[12]   
14                                         Barnet[12]   
15                                        Enfield[12]   
16                                     Wandsworth[13]   
17                                      Southwark[14]   
18                                           City[14]   
19                           Barking and Dagenham[14]   
20                                      Redbridge[15]   
21                                         Bexley[15]   
22                           Richmond upon Thames[15]   
23                                         Bexley[16]   
24                                             Barnet   
25                                         Barnet[16]   
26                                      Islington[17]   
27                                     Wandsworth[18]   
28                                    Westminster[19]   
29                                        Bromley[20]   
30                                         Newham[20]   
31                           Barking and Dagenham[20]   
32                           Barking and Dagenham[21]   
33                                         Sutton[21]   
34                                         Ealing[21]   
35                                    Westminster[22]   
36                                       Lewisham[22]   
37                                         Harrow[22]   
38                                         Sutton[22]   
39                                         Camden[23]   
40                                         Bexley[23]   
41                                      Southwark[24]   
42                           Kingston upon Thames[24]   
43                                  Tower Hamlets[25]   
44                                         Bexley[25]   
45                                         Bexley[26]   
46                                        Bromley[26]   
47                                        Bromley[26]   
48                                         Bexley[27]   
49                                           City[27]   
50                                       Lewisham[28]   
51                                          Greenwich   
52                                  Tower Hamlets[28]   
53                                             Bexley   
54                                         Camden[29]   
55                                        Enfield[30]   
56                                       Haringey[31]   
57                                  Tower Hamlets[31]   
58                                       Haringey[32]   
59                                       Hounslow[33]   
60                                             Barnet   
61                                              Brent   
62                                        Enfield[34]   
63                                        Lambeth[34]   
64                                       Lewisham[34]   
65                                        Bromley[35]   
66                                  Tower Hamlets[36]   
67                                        Bromley[35]   
68   Kensington and ChelseaHammersmith and Fulham[36]   
69                                          Brent[

Let's remove the Square brackets [ ] and numbers from the borough column

In [15]:
df1['Neighbourhood'] = df1['Neighbourhood'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

Neighbourhood  \
0                              Bexley, Greenwich    
1                  Ealing, Hammersmith and Fulham   
2                                         Croydon   
3                                         Croydon   
4                                          Bexley   
5                                       Redbridge   
6                                            City   
7                                     Westminster   
8                                           Brent   
9                                         Bromley   
10                                      Islington   
11                                        Bromley   
12                                      Islington   
13                                       Havering   
14                                         Barnet   
15                                        Enfield   
16                                     Wandsworth   
17                                      Southwark   
18                                           City   
19                           Barking and Dagenham   
20                                      Redbridge   
21                                         Bexley   
22                           Richmond upon Thames   
23                                         Bexley   
24                                         Barnet   
25                                         Barnet   
26                                      Islington   
27                                     Wandsworth   
28                                    Westminster   
29                                        Bromley   
30                                         Newham   
31                           Barking and Dagenham   
32                           Barking and Dagenham   
33                                         Sutton   
34                                         Ealing   
35                                    Westminster   
36                                       Lewisham   
37                                         Harrow   
38                                         Sutton   
39                                         Camden   
40                                         Bexley   
41                                      Southwark   
42                           Kingston upon Thames   
43                                  Tower Hamlets   
44                                         Bexley   
45                                         Bexley   
46                                        Bromley   
47                                        Bromley   
48                                         Bexley   
49                                           City   
50                                       Lewisham   
51                                      Greenwich   
52                                  Tower Hamlets   
53                                         Bexley   
54                                         Camden   
55                                        Enfield   
56                                       Haringey   
57                                  Tower Hamlets   
58                                       Haringey   
59                                       Hounslow   
60                                         Barnet   
61                                          Brent   
62                                        Enfield   
63                                        Lambeth   
64                                       Lewisham   
65                                        Bromley   
66                                  Tower Hamlets   
67                                        Bromley   
68   Kensington and ChelseaHammersmith and Fulham   
69                                          Brent   
70                                         Barnet   
71                                        Enfield   
72                                         Barnet   
73                                         Barnet   
74                                      Southwark   
75      

Take the dimension of the dataframe

In [16]:
df1.shape

(532, 3)

We currently have 533 records and 3 columns of our data. It's time to perform Feature Engineering

#### Feature Engineering

We can only focusing on the neighbourhoods of London, so performing the changes

In [18]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

Neighbourhood  \
0                              Bexley, Greenwich    
1                  Ealing, Hammersmith and Fulham   
6                                            City   
7                                     Westminster   
9                                         Bromley   
10                                      Islington   
12                                      Islington   
14                                         Barnet   
15                                        Enfield   
16                                     Wandsworth   
17                                      Southwark   
18                                           City   
22                           Richmond upon Thames   
24                                         Barnet   
26                                      Islington   
27                                     Wandsworth   
28                                    Westminster   
29                                        Bromley   
30                                         Newham   
34                                         Ealing   
35                                    Westminster   
36                                       Lewisham   
39                                         Camden   
41                                      Southwark   
43                                  Tower Hamlets   
45                                         Bexley   
49                                           City   
50                                       Lewisham   
51                                      Greenwich   
52                                  Tower Hamlets   
54                                         Camden   
56                                       Haringey   
57                                  Tower Hamlets   
58                                       Haringey   
60                                         Barnet   
61                                          Brent   
63                                        Lambeth   
64                                       Lewisham   
66                                  Tower Hamlets   
68   Kensington and ChelseaHammersmith and Fulham   
69                                          Brent   
70                                         Barnet   
73                                         Barnet   
74                                      Southwark   
75                                  Tower Hamlets   
76                                         Camden   
77                                  Tower Hamlets   
78                                 Waltham Forest   
79                                         Newham   
80                                      Islington   
82                           Richmond upon Thames   
84                                       Lewisham   
86                                         Camden   
87                                    Westminster   
88                                      Greenwich   
91                         Kensington and Chelsea   
94                                         Barnet   
95                                    Westminster   
96                                       Lewisham   
97                                 Waltham Forest   
99       Hounslow, Ealing, Hammersmith and Fulham   
100                                         Brent   
101                                        Barnet   
102                           Lambeth, Wandsworth   
103                                     Islington   
105                                        Barnet   
107                                        Merton   
108                                        Barnet   
114                                   Westminster   
121                         Barnet, Brent, Camden   
122                                      Lewisham   
124                                        Bexley   
125                                      Haringey   
127                                       Bromley   
128                                 Tower Hamlets   
130     

In [19]:
df1.shape

(309, 3)

We now have only 310 rows. We can proceed with our further steps. Getting some descriptive statistics

In [20]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 309 entries, 0 to 529
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Neighbourhood  309 non-null    object
 1   town           309 non-null    object
 2   post_code      309 non-null    object
dtypes: object(3)
memory usage: 9.7+ KB


## Geolocations of the London Neighbourhoods

#### ArcGis API

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so. 

In [21]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

Defining London arcgis geocode function to return latitude and longitude

In [22]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

Checking sample data

In [23]:
c = get_x_y_uk('SE2')
c

'51.492450000000076,0.12127000000003818'

Looks good, We Copy over the postal codes of london to pass it into the geolocator function that we just defined above

In [24]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0                      SE2
1                   W3, W4
6                      EC3
7                      WC2
9                     SE20
10                 EC1, N1
12                     N19
14                EN5, NW7
15                N11, N14
16                    SW12
17                     SE1
18                EC1, EC2
22                    SW13
24                NW7, EN5
26                      N1
27                    SW11
28                      W2
29               BR3, SE20
30           E6, E16, IG11
34                      W4
35                     SW1
36                     SE6
39                     NW3
41                     SE1
43                      E2
45           DA6, DA7, SE2
49                     EC4
50                     SE3
51               SE3, SE12
52                     E14
54                     WC1
56                N11, N22
57                      E3
58                     N22
60                NW2, NW4
61                    NW10
63           SW2, SW9, SE5
6

Passing postal codes of london to get the geographical co-ordinates

In [25]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0        51.492450000000076,0.12127000000003818
1         51.51324000000005,-0.2674599999999714
6        51.51200000000006,-0.08057999999994081
7        51.51651000000004,-0.11967999999995982
9        51.41009000000008,-0.05682999999993399
10      51.523610000000076,-0.09876999999994496
12       51.56393000000003,-0.12944999999996298
14       51.615680000000054,-0.2451099999999542
15       51.63429000000008,-0.13365999999996347
16        51.44822000000005,-0.1483899999999494
17      51.499960000000044,-0.09567999999995891
18      51.518410000000074,-0.08814999999992779
22       51.47469000000007,-0.24162999999992962
24       51.615680000000054,-0.2451099999999542
26       51.52969000000007,-0.08696999999995114
27       51.46760000000006,-0.16289999999997917
28        51.51494000000008,-0.1804799999999318
29       51.41009000000008,-0.05682999999993399
30       51.50913000000003,0.015280000000075233
34       51.48944000000006,-0.26193999999992457
35       51.49713000000003,-0.1382899999

#### Latitude

Extracting the latitude from our previously collected coordinates

In [26]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
10     51.523610000000076
12      51.56393000000003
14     51.615680000000054
15      51.63429000000008
16      51.44822000000005
17     51.499960000000044
18     51.518410000000074
22      51.47469000000007
24     51.615680000000054
26      51.52969000000007
27      51.46760000000006
28      51.51494000000008
29      51.41009000000008
30      51.50913000000003
34      51.48944000000006
35      51.49713000000003
36      51.43722000000008
39     51.555060000000026
41     51.499960000000044
43      51.52669000000003
45     51.492450000000076
49      51.51389000000006
50      51.47138000000007
51      51.44759000000005
52      51.51122000000004
54     51.524500000000046
56      51.60214000000008
57      51.52702000000005
58      51.60214000000008
60     51.589180000000056
61      51.53938000000005
63      51.47480000000007
64     51.462680000000034
66      51.5

#### Longitude

Extracting the Longitude from our previously collected coordinates

In [27]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0        0.12127000000003818
1        -0.2674599999999714
6       -0.08057999999994081
7       -0.11967999999995982
9       -0.05682999999993399
10      -0.09876999999994496
12      -0.12944999999996298
14       -0.2451099999999542
15      -0.13365999999996347
16       -0.1483899999999494
17      -0.09567999999995891
18      -0.08814999999992779
22      -0.24162999999992962
24       -0.2451099999999542
26      -0.08696999999995114
27      -0.16289999999997917
28       -0.1804799999999318
29      -0.05682999999993399
30      0.015280000000075233
34      -0.26193999999992457
35      -0.13828999999992675
36     -0.018679999999960728
39       -0.1734799999999268
41      -0.09567999999995891
43      -0.06256999999993695
45       0.12127000000003818
49      -0.10433999999997923
50      0.023380000000031487
51      0.013500000000021828
52     -0.012639999999976226
54      -0.12272999999993317
56      -0.11222999999995409
57      -0.02593999999993457
58      -0.11222999999995409
60      -0.228

We now have the geographical co-ordinates of the London Neighbourhoods.

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage

In [28]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['Neighbourhood','town','post_code','latitude','longitude']
london_merged

Neighbourhood  \
0                              Bexley, Greenwich    
1                  Ealing, Hammersmith and Fulham   
6                                            City   
7                                     Westminster   
9                                         Bromley   
10                                      Islington   
12                                      Islington   
14                                         Barnet   
15                                        Enfield   
16                                     Wandsworth   
17                                      Southwark   
18                                           City   
22                           Richmond upon Thames   
24                                         Barnet   
26                                      Islington   
27                                     Wandsworth   
28                                    Westminster   
29                                        Bromley   
30                                         Newham   
34                                         Ealing   
35                                    Westminster   
36                                       Lewisham   
39                                         Camden   
41                                      Southwark   
43                                  Tower Hamlets   
45                                         Bexley   
49                                           City   
50                                       Lewisham   
51                                      Greenwich   
52                                  Tower Hamlets   
54                                         Camden   
56                                       Haringey   
57                                  Tower Hamlets   
58                                       Haringey   
60                                         Barnet   
61                                          Brent   
63                                        Lambeth   
64                                       Lewisham   
66                                  Tower Hamlets   
68   Kensington and ChelseaHammersmith and Fulham   
69                                          Brent   
70                                         Barnet   
73                                         Barnet   
74                                      Southwark   
75                                  Tower Hamlets   
76                                         Camden   
77                                  Tower Hamlets   
78                                 Waltham Forest   
79                                         Newham   
80                                      Islington   
82                           Richmond upon Thames   
84                                       Lewisham   
86                                         Camden   
87                                    Westminster   
88                                      Greenwich   
91                         Kensington and Chelsea   
94                                         Barnet   
95                                    Westminster   
96                                       Lewisham   
97                                 Waltham Forest   
99       Hounslow, Ealing, Hammersmith and Fulham   
100                                         Brent   
101                                        Barnet   
102                           Lambeth, Wandsworth   
103                                     Islington   
105                                        Barnet   
107                                        Merton   
108                                        Barnet   
114                                   Westminster   
121                         Barnet, Brent, Camden   
122                                      Lewisham   
124                                        Bexley   
125                                      Haringey   
127                                       Bromley   
128                                 Tower Hamlets   
130     

In [29]:
london_merged.dtypes

Neighbourhood     object
town              object
post_code         object
latitude         float64
longitude        float64
dtype: object

#### Co-ordinates for London

Getting the geocode for London to help visualize it on the map

In [31]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [32]:
london_lat_coords

51.50642000000005

#### Visualize the Map of London

To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [33]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, Neighbourhood, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['Neighbourhood'], london_merged['town']):
    label = '{}, {}'.format(town, Neighbourhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

#### Venues in London

To proceed with the next part, we need to define Foursquare API credentials.

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in London.

In [51]:
CLIENT_ID = '3XW3X3GAEY00VJGDSZXKRT5QWYOPBZFOFPQ1SKL2VBKBVD0G' 
CLIENT_SECRET = 'SLIM5FWPD2TSJD3P2QPV0QYYKX23OJPHHCUXVQRLZVBT2CNH'
VERSION = '20180605' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis

In [52]:
LIMIT=100
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [53]:
venues_in_London = getNearbyVenues(names=london_merged['Neighbourhood'], latitudes=london_merged['latitude'], longitudes=london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and ChelseaHammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames
En

Sampling our data

In [54]:
venues_in_London.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue  Venue Latitude  Venue Longitude  \
0                      Lesnes Abbey       51.489526         0.125839   
1                       Sainsbury's       51.492826         0.120524   
2                              Lidl       51.496152         0.118417   
3  Abbey Wood Railway Station (ABW)       51.490825         0.123432   
4                       Bean @ Work       51.491172         0.120649   

  Venue Category  
0  Historic Site  
1    Supermarket  
2    Supermarket  
3  Train Station  
4    Coffee Shop

In [55]:
venues_in_London.shape

(10521, 7)

Wow, we have scraped together 10521 records for venues. This will definitely make the clustering interesting.

In [96]:
print('There are {} Uniques Categories.'.format(len(venues_in_London['Venue Category'].unique())))
venues_in_London.groupby('Neighbourhood').count().head()

There are 292 Uniques Categories.


Neighbourhood Latitude  Neighbourhood Longitude  Venue  \
Neighbourhood                                                                   
Barnet                                    562                      562    562   
Barnet, Brent, Camden                       5                        5      5   
Bexley                                     23                       23     23   
Bexley, Greenwich                           7                        7      7   
Bexley, Greenwich                           6                        6      6   

                       Venue Latitude  Venue Longitude  Venue Category  
Neighbourhood                                                           
Barnet                            562              562             562  
Barnet, Brent, Camden               5                5               5  
Bexley                             23               23              23  
Bexley, Greenwich                   7                7               7  
Bexley, Greenwich                   6                6               6

#### Grouping by Venue Categories
We need to now see how many Venue Categories are there for further processing

In [97]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                                           
Adult Boutique                                                               Islington   
African Restaurant                                                         Westminster   
American Restaurant                                                     Waltham Forest   
Antique Shop                                                               Westminster   
Arcade                                                                     Westminster   
Arepa Restaurant                                                         Tower Hamlets   
Argentinian Restaurant                                                     Westminster   
Art Gallery                                                                Westminster   
Art Museum                                                   Redbridge, Waltham Forest   
Arts & Crafts Store                                                        Westminster   
Asian Restaurant                                                           Westminster   
Athletics & Sports                                                           Southwark   
Australian Restaurant                                                      Westminster   
Auto Garage                                                                  Islington   
Auto Workshop                                                                    Brent   
BBQ Joint                                                                    Redbridge   
Bagel Shop                                                                 Westminster   
Bakery                                                                     Westminster   
Bar                                                                        Westminster   
Beach                                                                           Newham   
Bed & Breakfast                                                            Westminster   
Beer Bar                                                                   Westminster   
Beer Garden                                                                  Southwark   
Beer Store                                                                  Wandsworth   
Bike Shop                                                                    Southwark   
Bistro                                                                       Southwark   
Boarding House                                                  Hammersmith and Fulham   
Bookstore                                                                  Westminster   
Botanical Garden                                                                  City   
Boutique                                                                   Westminster   
Bowling Alley                                                              Westminster   
Boxing Gym                                Kensington and ChelseaHammersmith and Fulham   
Brasserie                                                                  Westminster   
Brazilian Restaurant                                                        Wandsworth   
Breakfast Spot                                                             Westminster   
Brewery                                                                  Tower Hamlets   
Building                                                             City, Westminster   
Burger Joint                                                               Westminster   
Burrito Place                                                              Westminster   
Bus Station                                                                 Wandsworth   
Bus Stop                                                                   Westminster   
Butcher                                                                        Hackney   
Café                                                                       Westminster   
Camera Store           

We can see 292 records, just goes to show how diverse and interesting the place is.

#### One Hot Encoding 
We need to Encode our venue categories to get a better result for our clustering

In [98]:
# one hot encoding
London_onehot = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
London_onehot['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [London_onehot.columns[-1]] + list(London_onehot.columns[:-1])
London_onehot = London_onehot[fixed_columns]
London_grouped = London_onehot.groupby('Neighbourhood').mean().reset_index()
London_onehot.head(5)

Neighbourhood  Adult Boutique  African Restaurant  \
0  Bexley, Greenwich                0                   0   
1  Bexley, Greenwich                0                   0   
2  Bexley, Greenwich                0                   0   
3  Bexley, Greenwich                0                   0   
4  Bexley, Greenwich                0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                    0             0       0                 0   
1                    0             0       0                 0   
2                    0             0       0                 0   
3                    0             0       0                 0   
4                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  Auto Garage  \
0                 0                   0                      0            0   
1                 0                   0                      0            0   
2                 0                   0                      0            0   
3                 0                   0                      0            0   
4                 0                   0                      0            0   

   Auto Workshop  BBQ Joint  Bagel Shop  Bakery  Bar  Beach  Bed & Breakfast  \
0              0          0           0       0    0      0                0   
1              0          0           0       0    0      0                0   
2              0          0           0       0    0      0                0   
3              0          0           0       0    0      0                0   
4              0          0           0       0    0      0                0   

   Beer Bar  Beer Garden  Beer Store  Bike Shop  Bistro  Boarding House  \
0         0            0           0          0       0               0   
1         0            0           0          0       0               0   
2         0            0           0          0       0               0   
3         0            0           0          0       0               0   
4         0            0           0          0       0               0   

   Bookstore  Botanical Garden  Boutique  Bowling Alley  Boxing Gym  \
0          0                 0         0              0           0   
1          0                 0         0              0           0   
2          0                 0         0              0           0   
3          0                 0         0              0           0   
4          0                 0         0              0           0   

   Brasserie  Brazilian Restaurant  Breakfast Spot  Brewery  Building  \
0          0                     0               0        0         0   
1          0                     0               0        0         0   
2          0                     0               0        0         0   
3          0                     0               0        0         0   
4          0                     0               0        0         0   

   Burger Joint  Burrito Place  Bus Station  Bus Stop  Butcher  Café  \
0             0              0            0         0        0     0   
1             0              0            0         0        0     0   
2             0              0            0         0        0     0   
3             0              0            0         0        0     0   
4             0              0            0         0        0     0   

   Camera Store  Canal  Candy Store  Caribbean Restaurant  \
0             0      0    

Grouping of each Neighbourhoods with 10 common venues:

In [99]:
num_top_venues = 5
for hood in London_grouped['Neighbourhood']:
    print("---- "+hood+" ----")
    temp =London_grouped[London_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Barnet ----
                venue  freq
0         Coffee Shop  0.09
1                Café  0.07
2       Grocery Store  0.06
3  Italian Restaurant  0.04
4                 Pub  0.04


---- Barnet, Brent, Camden ----
                  venue  freq
0        Clothing Store   0.2
1                Bakery   0.2
2           Supermarket   0.2
3  Gym / Fitness Center   0.2
4           Bus Station   0.2


---- Bexley ----
               venue  freq
0        Supermarket  0.26
1      Historic Site  0.17
2        Coffee Shop  0.13
3  Convenience Store  0.13
4      Train Station  0.13


---- Bexley, Greenwich ----
                        venue  freq
0                    Bus Stop  0.14
1  Construction & Landscaping  0.14
2                 Sports Club  0.14
3                        Park  0.14
4                 Golf Course  0.14


---- Bexley, Greenwich  ----
               venue  freq
0        Supermarket  0.33
1        Coffee Shop  0.17
2      Train Station  0.17
3  Convenience Store  0.17
4      H

Creating new dataframe — Putting the common venues into pandas dataframe, the following return_most_common_venuesis used to sort the venues in descending order.


In [100]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Then we create a new panda dataframe with 10 most common venues

In [101]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden           Bus Station                Bakery   
2                 Bexley           Supermarket         Historic Site   
3      Bexley, Greenwich           Sports Club     Convenience Store   
4     Bexley, Greenwich            Supermarket         Historic Site   

  3rd Most Common Venue 4th Most Common Venue       5th Most Common Venue  \
0         Grocery Store    Italian Restaurant                         Pub   
1  Gym / Fitness Center        Clothing Store                 Supermarket   
2           Coffee Shop     Convenience Store               Train Station   
3           Golf Course                  Park  Construction & Landscaping   
4         Train Station     Convenience Store                 Coffee Shop   

  6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
0              Bus Stop           Supermarket            Sushi Restaurant   
1           Zoo Exhibit     Fish & Chips Shop              Farmers Market   
2                  Park              Bus Stop  Construction & Landscaping   
3         Historic Site              Bus Stop           Food & Drink Shop   
4           Fish Market        Farmers Market        Fast Food Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0              Pharmacy     Turkish Restaurant  
1  Fast Food Restaurant    Filipino Restaurant  
2           Golf Course                Exhibit  
3            Food Court            Flower Shop  
4   Filipino Restaurant            Film Studio

#### Model Building

#### K Means
Let's cluster the city of london to roughly 5 to make it easier to analyze. 

We use the K Means clustering technique to do so.

In [102]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(n_clusters=5, random_state=0)

#### Labelling Clustered Data

In [103]:
kmeans_london.labels_[0:10]

array([0, 4, 3, 1, 3, 0, 0, 0, 0, 0], dtype=int32)

In [104]:
# Creating a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
London_merged = London_grouped.copy()
# Add the clustering labels
London_merged["Cluster Labels"] = kmeans_london.labels_
London_merged.rename(columns={"borough": "Neighbourhood"}, inplace=True)
London_merged.head(10)

Neighbourhood  Adult Boutique  African Restaurant  \
0                 Barnet             0.0                 0.0   
1  Barnet, Brent, Camden             0.0                 0.0   
2                 Bexley             0.0                 0.0   
3      Bexley, Greenwich             0.0                 0.0   
4     Bexley, Greenwich              0.0                 0.0   
5                  Brent             0.0                 0.0   
6          Brent, Camden             0.0                 0.0   
7          Brent, Ealing             0.0                 0.0   
8          Brent, Harrow             0.0                 0.0   
9                Bromley             0.0                 0.0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0             0.007117           0.0     0.0               0.0   
1             0.000000           0.0     0.0               0.0   
2             0.000000           0.0     0.0               0.0   
3             0.000000           0.0     0.0               0.0   
4             0.000000           0.0     0.0               0.0   
5             0.000000           0.0     0.0               0.0   
6             0.000000           0.0     0.0               0.0   
7             0.000000           0.0     0.0               0.0   
8             0.000000           0.0     0.0               0.0   
9             0.000000           0.0     0.0               0.0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                0.007117     0.000000    0.000000                  0.0   
1                0.000000     0.000000    0.000000                  0.0   
2                0.000000     0.000000    0.000000                  0.0   
3                0.000000     0.000000    0.000000                  0.0   
4                0.000000     0.000000    0.000000                  0.0   
5                0.000000     0.000000    0.000000                  0.0   
6                0.000000     0.000000    0.000000                  0.0   
7                0.000000     0.000000    0.000000                  0.0   
8                0.000000     0.034091    0.022727                  0.0   
9                0.000000     0.000000    0.000000                  0.0   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  Auto Garage  \
0          0.019573                 0.0                    0.0          0.0   
1          0.000000                 0.0                    0.0          0.0   
2          0.000000                 0.0                    0.0          0.0   
3          0.000000                 0.0                    0.0          0.0   
4          0.000000                 0.0                    0.0          0.0   
5          0.013514                 0.0                    0.0          0.0   
6          0.000000                 0.0                    0.0          0.0   
7          0.000000                 0.0                    0.0          0.0   
8          0.000000                 0.0                    0.0          0.0   
9          0.000000                 0.0                    0.0          0.0   

   Auto Workshop  BBQ Joint  Bagel Shop    Bakery       Bar  Beach  \
0       0.007117   0.001779    0.005338  0.016014  0.005338    0.0   
1       0.000000   0.000000    0.000000  0.200000  0.000000    0.0   
2       0.000000   0.000000    0.000000  0.000000  0.000000    0.0   
3       0.000000   0.000000    0.000000  0.000000  0.000000    0.0   
4       0.000000   0.000000    0.000000  0.000000  0.000000    0.0   
5       0.006757   0.000000    0.000000  0.006757  0.000000    0.0   
6       0.000000   0.000000    0.000000  0.000000  0.000000    0.0   
7       0.000000   0.000000    0.000000  0.000000  0.000000    0.0   
8       0.000000   0.000000    0.000000  0.011364  0.011364    0.0   
9       0.000000   0.000000    0.000000  0.000000  0.000000    0.0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Bike Shop    Bistro  \
0              0.0  0.005338      

In [105]:
London_merged['Latitude'] = venues_in_London['Neighbourhood Latitude']
London_merged['Longitude'] = venues_in_London['Neighbourhood Longitude']
# Sorting the results by Cluster Labels
London_merged.sort_values(["Cluster Labels"], inplace=True)
London_merged

Neighbourhood  Adult Boutique  \
0                                         Barnet        0.000000   
27      Hounslow, Ealing, Hammersmith and Fulham        0.000000   
28                                     Islington        0.007102   
29                              Islington & City        0.000000   
30                             Islington, Camden        0.000000   
31                        Kensington and Chelsea        0.000000   
32  Kensington and ChelseaHammersmith and Fulham        0.000000   
33                          Kingston upon Thames        0.000000   
34                                       Lambeth        0.000000   
35                            Lambeth, Southwark        0.000000   
36                           Lambeth, Wandsworth        0.000000   
37                                      Lewisham        0.000000   
38                             Lewisham, Bromley        0.000000   
39                           Lewisham, Southwark        0.000000   
40                                        Merton        0.000000   
41                                        Newham        0.000000   
42                                     Redbridge        0.000000   
43                     Redbridge, Waltham Forest        0.000000   
44                          Richmond upon Thames        0.000000   
45                                     Southwark        0.000000   
46                                 Tower Hamlets        0.000000   
47                                Waltham Forest        0.000000   
26                                      Hounslow        0.000000   
48                                    Wandsworth        0.000000   
24                           Haringey, Islington        0.000000   
11                          Camden and Islington        0.000000   
5                                          Brent        0.000000   
6                                  Brent, Camden        0.000000   
7                                  Brent, Ealing        0.000000   
8                                  Brent, Harrow        0.000000   
9                                        Bromley        0.000000   
10                                        Camden        0.000000   
23                           Haringey and Barnet        0.000000   
12                                          City        0.000000   
13                             City, Westminster        0.000000   
49                                   Westminster        0.000000   
15                                        Ealing        0.000000   
17                                       Enfield        0.000000   
18                                     Greenwich        0.000000   
19                           Greenwich, Lewisham        0.000000   
20                                       Hackney        0.003378   
21                        Hammersmith and Fulham        0.000000   
22                                      Haringey        0.000000   
14                                       Croydon        0.000000   
16                Ealing, Hammersmith and Fulham        0.000000   
3                              Bexley, Greenwich        0.000000   
25                                 Harrow, Brent        0.000000   
4                             Bexley, Greenwich         0.000000   
2                                         Bexley        0.000000   
1                          Barnet, Brent, Camden        0.000000   

    African Restaurant  American Restaurant  Antique Shop    Arcade  \
0             0.000000             0.007117      0.000000  0.000000   
27            0.000000             0.000000      0.000000  0.000000   
28            0.000000             0.000000      0.000000  0.000000   
29            0.000000             0.000000      0.000000  0.000000   
30            0.000000             0.000000      0.000000  0.000000   
31            0.000000             0.000000      0.005217  0.000000   
32            0.000000             0.000000      0.000000  0.000000  

Drop all the NaN values to prevent data skew

In [106]:
London_merged_nonan = London_merged.dropna(subset=['Cluster Labels'])

#### Visualizing the clustered neighbourhood
Let's plot the clusters

In [109]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(London_merged_nonan['Latitude'], London_merged_nonan['Longitude'], London_merged_nonan['Neighbourhood'], London_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

In [110]:
# create map
map_clusters = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=11)
# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(London_merged_nonan['Latitude'], London_merged_nonan['Longitude'], London_merged_nonan['Neighbourhood'], London_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=20,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
display(map_clusters)

#### Examining our Clusters

In [72]:
London_merged_nonan.loc[London_merged_nonan['Cluster Labels'] == 1, London_merged_nonan.columns[[1] + list(range(5, London_merged_nonan.shape[1]))]]

Adult Boutique  Arcade  Arepa Restaurant  Argentinian Restaurant  \
16             0.0     0.0               0.0                     0.0   
3              0.0     0.0               0.0                     0.0   

    Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
16          0.0         0.0                  0.0               0.0   
3           0.0         0.0                  0.0               0.0   

    Athletics & Sports  Australian Restaurant  Auto Garage  Auto Workshop  \
16                 0.0                    0.0          0.0            0.0   
3                  0.0                    0.0          0.0            0.0   

    BBQ Joint  Bagel Shop  Bakery  Bar  Beach  Bed & Breakfast  Beer Bar  \
16        0.0         0.0     0.0  0.0    0.0              0.0       0.0   
3         0.0         0.0     0.0  0.0    0.0              0.0       0.0   

    Beer Garden  Beer Store  Bike Shop  Bistro  Boarding House  Bookstore  \
16          0.0         0.0        0.0     0.0             0.0        0.0   
3           0.0         0.0        0.0     0.0             0.0        0.0   

    Botanical Garden  Boutique  Bowling Alley  Boxing Gym  Brasserie  \
16               0.0       0.0            0.0         0.0        0.0   
3                0.0       0.0            0.0         0.0        0.0   

    Brazilian Restaurant  Breakfast Spot  Brewery  Building  Burger Joint  \
16                   0.0        0.166667      0.0       0.0           0.0   
3                    0.0        0.000000      0.0       0.0           0.0   

    Burrito Place  Bus Station  Bus Stop  Butcher  Café  Camera Store  Canal  \
16            0.0          0.0  0.000000      0.0   0.0           0.0    0.0   
3             0.0          0.0  0.142857      0.0   0.0           0.0    0.0   

    Candy Store  Caribbean Restaurant  Caucasian Restaurant  Cemetery  \
16          0.0                   0.0                   0.0       0.0   
3           0.0                   0.0                   0.0       0.0   

    Champagne Bar  Cheese Shop  Child Care Service  Chinese Restaurant  \
16            0.0          0.0                 0.0                 0.0   
3             0.0          0.0                 0.0                 0.0   

    Chocolate Shop  Church  Circus School  Climbing Gym  Clothing Store  \
16             0.0     0.0            0.0           0.0             0.0   
3              0.0     0.0            0.0           0.0             0.0   

    Cocktail Bar  Coffee Shop  College Quad  Comedy Club  \
16           0.0          0.0           0.0          0.0   
3            0.0          0.0           0.0          0.0   

    Comfort Food Restaurant  Community Center  Concert Hall  \
16                      0.0               0.0           0.0   
3                       0.0               0.0           0.0   

    Construction & Landscaping  Convenience Store  Convention Center  \
16                    0.000000           0.000000                0.0   
3                     0.142857           0.142857                0.0   

    Cosmetics Shop  Costume Shop  Creperie  Cricket Ground  Cupcake Shop  \
16             0.0           0.0       0.0             0.0           0.0   
3              0.0           0.0       0.0             0.0           0.0   

    Cycle Studio  Dance Studio  Deli / Bodega  Department Store  Dessert Shop  \
16           0.0           0.0            0.0               0.0           0.0   
3            0.0           0.0            0.0               0.0           0.0   

    Dim Sum Restaurant  Diner  Discount Store  Distillery  Dog Run  \
16                 0.0    0.0             0.0         0.0      0.0   
3                  0.0    0.0             0.0         0.0      0.0   

    Doner Restaurant  Donut Shop  Dry Cleaner  Electronics Store  \
16               0.0         0.0          0.0                0.0   
3                0.0         0.0          0.0                0.0   

    English Restaurant  Escape Room  Ethiopian 

Cluster 2

In [73]:
London_merged_nonan.loc[London_merged_nonan['Cluster Labels'] == 2, London_merged_nonan.columns[[1] + list(range(5, London_merged_nonan.shape[1]))]]

Adult Boutique  Arcade  Arepa Restaurant  Argentinian Restaurant  \
25             0.0     0.0               0.0                     0.0   

    Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
25          0.0         0.0                  0.0               0.0   

    Athletics & Sports  Australian Restaurant  Auto Garage  Auto Workshop  \
25                 0.0                    0.0          0.0            0.0   

    BBQ Joint  Bagel Shop    Bakery  Bar  Beach  Bed & Breakfast  Beer Bar  \
25        0.0         0.0  0.333333  0.0    0.0              0.0       0.0   

    Beer Garden  Beer Store  Bike Shop  Bistro  Boarding House  Bookstore  \
25          0.0         0.0        0.0     0.0             0.0        0.0   

    Botanical Garden  Boutique  Bowling Alley  Boxing Gym  Brasserie  \
25               0.0       0.0            0.0         0.0        0.0   

    Brazilian Restaurant  Breakfast Spot  Brewery  Building  Burger Joint  \
25                   0.0             0.0      0.0       0.0           0.0   

    Burrito Place  Bus Station  Bus Stop  Butcher  Café  Camera Store  Canal  \
25            0.0          0.0       0.0      0.0   0.0           0.0    0.0   

    Candy Store  Caribbean Restaurant  Caucasian Restaurant  Cemetery  \
25          0.0                   0.0                   0.0       0.0   

    Champagne Bar  Cheese Shop  Child Care Service  Chinese Restaurant  \
25            0.0          0.0                 0.0                 0.0   

    Chocolate Shop  Church  Circus School  Climbing Gym  Clothing Store  \
25             0.0     0.0            0.0           0.0             0.0   

    Cocktail Bar  Coffee Shop  College Quad  Comedy Club  \
25           0.0          0.0           0.0          0.0   

    Comfort Food Restaurant  Community Center  Concert Hall  \
25                      0.0               0.0           0.0   

    Construction & Landscaping  Convenience Store  Convention Center  \
25                         0.0                0.0                0.0   

    Cosmetics Shop  Costume Shop  Creperie  Cricket Ground  Cupcake Shop  \
25             0.0           0.0       0.0             0.0           0.0   

    Cycle Studio  Dance Studio  Deli / Bodega  Department Store  Dessert Shop  \
25           0.0           0.0            0.0               0.0           0.0   

    Dim Sum Restaurant  Diner  Discount Store  Distillery  Dog Run  \
25                 0.0    0.0             0.0         0.0      0.0   

    Doner Restaurant  Donut Shop  Dry Cleaner  Electronics Store  \
25               0.0         0.0          0.0                0.0   

    English Restaurant  Escape Room  Ethiopian Restaurant  Event Space  \
25                 0.0          0.0                   0.0          0.0   

    Exhibit  Falafel Restaurant  Farmers Market  Fast Food Restaurant  \
25      0.0                 0.0             0.0                   0.0   

    Filipino Restaurant  Film Studio  Fish & Chips Shop  Fish Market  \
25                  0.0          0.0                0.0          0.0   

    Fishing Store  Flea Market  Flower Shop  Food & Drink Shop  Food Court  \
25            0.0          0.0          0.0                0.0         0.0   

    Food Stand  Food Truck  Fountain  French Restaurant  Fried Chicken Joint  \
25         0.0         0.0       0.0                0.0                  0.0   

    Fruit & Vegetable Store  Furniture / Home Store  Gaming Cafe  Garden  \
25                      0.0                     0.0          0.0     0.0   

    Garden Center  Gas Station  Gastropub  Gay Bar  Gelato Shop  \
25            0.0          0.0        0.0      0.0          0.0   

    General Entertainment  German Restaurant  Gift Shop  Golf Course  \
25                    0.0                0.0        0.0          0.0   

    Gourmet Shop  Greek Restaurant  Grilled Meat Restaurant  Grocery Store  \
25           0.0               0.0                      0.0            0.0   

       

Cluster 3

In [74]:
London_merged_nonan.loc[London_merged_nonan['Cluster Labels'] == 3, London_merged_nonan.columns[[1] + list(range(5, London_merged_nonan.shape[1]))]]

Adult Boutique  Arcade  Arepa Restaurant  Argentinian Restaurant  \
4             0.0     0.0               0.0                     0.0   
2             0.0     0.0               0.0                     0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
4          0.0         0.0                  0.0               0.0   
2          0.0         0.0                  0.0               0.0   

   Athletics & Sports  Australian Restaurant  Auto Garage  Auto Workshop  \
4                 0.0                    0.0          0.0            0.0   
2                 0.0                    0.0          0.0            0.0   

   BBQ Joint  Bagel Shop  Bakery  Bar  Beach  Bed & Breakfast  Beer Bar  \
4        0.0         0.0     0.0  0.0    0.0              0.0       0.0   
2        0.0         0.0     0.0  0.0    0.0              0.0       0.0   

   Beer Garden  Beer Store  Bike Shop  Bistro  Boarding House  Bookstore  \
4          0.0         0.0        0.0     0.0             0.0        0.0   
2          0.0         0.0        0.0     0.0             0.0        0.0   

   Botanical Garden  Boutique  Bowling Alley  Boxing Gym  Brasserie  \
4               0.0       0.0            0.0         0.0        0.0   
2               0.0       0.0            0.0         0.0        0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Building  Burger Joint  \
4                   0.0             0.0      0.0       0.0           0.0   
2                   0.0             0.0      0.0       0.0           0.0   

   Burrito Place  Bus Station  Bus Stop  Butcher  Café  Camera Store  Canal  \
4            0.0          0.0  0.000000      0.0   0.0           0.0    0.0   
2            0.0          0.0  0.043478      0.0   0.0           0.0    0.0   

   Candy Store  Caribbean Restaurant  Caucasian Restaurant  Cemetery  \
4          0.0                   0.0                   0.0       0.0   
2          0.0                   0.0                   0.0       0.0   

   Champagne Bar  Cheese Shop  Child Care Service  Chinese Restaurant  \
4            0.0          0.0                 0.0                 0.0   
2            0.0          0.0                 0.0                 0.0   

   Chocolate Shop  Church  Circus School  Climbing Gym  Clothing Store  \
4             0.0     0.0            0.0           0.0             0.0   
2             0.0     0.0            0.0           0.0             0.0   

   Cocktail Bar  Coffee Shop  College Quad  Comedy Club  \
4           0.0     0.166667           0.0          0.0   
2           0.0     0.130435           0.0          0.0   

   Comfort Food Restaurant  Community Center  Concert Hall  \
4                      0.0               0.0           0.0   
2                      0.0               0.0           0.0   

   Construction & Landscaping  Convenience Store  Convention Center  \
4                    0.000000           0.166667                0.0   
2                    0.043478           0.130435                0.0   

   Cosmetics Shop  Costume Shop  Creperie  Cricket Ground  Cupcake Shop  \
4             0.0           0.0       0.0             0.0           0.0   
2             0.0           0.0       0.0             0.0           0.0   

   Cycle Studio  Dance Studio  Deli / Bodega  Department Store  Dessert Shop  \
4           0.0           0.0            0.0               0.0           0.0   
2           0.0           0.0            0.0               0.0           0.0   

   Dim Sum Restaurant  Diner  Discount Store  Distillery  Dog Run  \
4                 0.0    0.0             0.0         0.0      0.0   
2                 0.0    0.0             0.0         0.0      0.0   

   Doner Restaurant  Donut Shop  Dry Cleaner  Electronics Store  \
4               0.0         0.0          0.0                0.0   
2               0.0         0.0          0.0                0.0   

   English Restaurant  Escape Room  Ethiopian Restaurant  Event Space  \
4                 0.0      

Cluster 4

In [75]:
London_merged_nonan.loc[London_merged_nonan['Cluster Labels'] == 4, London_merged_nonan.columns[[1] + list(range(5, London_merged_nonan.shape[1]))]]

Adult Boutique  Arcade  Arepa Restaurant  Argentinian Restaurant  \
1             0.0     0.0               0.0                     0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
1          0.0         0.0                  0.0               0.0   

   Athletics & Sports  Australian Restaurant  Auto Garage  Auto Workshop  \
1                 0.0                    0.0          0.0            0.0   

   BBQ Joint  Bagel Shop  Bakery  Bar  Beach  Bed & Breakfast  Beer Bar  \
1        0.0         0.0     0.2  0.0    0.0              0.0       0.0   

   Beer Garden  Beer Store  Bike Shop  Bistro  Boarding House  Bookstore  \
1          0.0         0.0        0.0     0.0             0.0        0.0   

   Botanical Garden  Boutique  Bowling Alley  Boxing Gym  Brasserie  \
1               0.0       0.0            0.0         0.0        0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Building  Burger Joint  \
1                   0.0             0.0      0.0       0.0           0.0   

   Burrito Place  Bus Station  Bus Stop  Butcher  Café  Camera Store  Canal  \
1            0.0          0.2       0.0      0.0   0.0           0.0    0.0   

   Candy Store  Caribbean Restaurant  Caucasian Restaurant  Cemetery  \
1          0.0                   0.0                   0.0       0.0   

   Champagne Bar  Cheese Shop  Child Care Service  Chinese Restaurant  \
1            0.0          0.0                 0.0                 0.0   

   Chocolate Shop  Church  Circus School  Climbing Gym  Clothing Store  \
1             0.0     0.0            0.0           0.0             0.2   

   Cocktail Bar  Coffee Shop  College Quad  Comedy Club  \
1           0.0          0.0           0.0          0.0   

   Comfort Food Restaurant  Community Center  Concert Hall  \
1                      0.0               0.0           0.0   

   Construction & Landscaping  Convenience Store  Convention Center  \
1                         0.0                0.0                0.0   

   Cosmetics Shop  Costume Shop  Creperie  Cricket Ground  Cupcake Shop  \
1             0.0           0.0       0.0             0.0           0.0   

   Cycle Studio  Dance Studio  Deli / Bodega  Department Store  Dessert Shop  \
1           0.0           0.0            0.0               0.0           0.0   

   Dim Sum Restaurant  Diner  Discount Store  Distillery  Dog Run  \
1                 0.0    0.0             0.0         0.0      0.0   

   Doner Restaurant  Donut Shop  Dry Cleaner  Electronics Store  \
1               0.0         0.0          0.0                0.0   

   English Restaurant  Escape Room  Ethiopian Restaurant  Event Space  \
1                 0.0          0.0                   0.0          0.0   

   Exhibit  Falafel Restaurant  Farmers Market  Fast Food Restaurant  \
1      0.0                 0.0             0.0                   0.0   

   Filipino Restaurant  Film Studio  Fish & Chips Shop  Fish Market  \
1                  0.0          0.0                0.0          0.0   

   Fishing Store  Flea Market  Flower Shop  Food & Drink Shop  Food Court  \
1            0.0          0.0          0.0                0.0         0.0   

   Food Stand  Food Truck  Fountain  French Restaurant  Fried Chicken Joint  \
1         0.0         0.0       0.0                0.0                  0.0   

   Fruit & Vegetable Store  Furniture / Home Store  Gaming Cafe  Garden  \
1                      0.0                     0.0          0.0     0.0   

   Garden Center  Gas Station  Gastropub  Gay Bar  Gelato Shop  \
1            0.0          0.0        0.0      0.0          0.0   

   General Entertainment  German Restaurant  Gift Shop  Golf Course  \
1                    0.0                0.0        0.0          0.0   

   Gourmet Shop  Greek Restaurant  Grilled Meat Restaurant  Grocery Store  \
1           0.0               0.0                      0.0            0.0   

   Gym  Gym / Fitness Center  Gym Pool  Halal Restaurant  Hardwa

Cluster 5

In [76]:
London_merged_nonan.loc[London_merged_nonan['Cluster Labels'] == 5, London_merged_nonan.columns[[1] + list(range(5, London_merged_nonan.shape[1]))]]

Empty DataFrame
Columns: [Adult Boutique, Arcade, Arepa Restaurant, Argentinian Restaurant, Art Gallery, Art Museum, Arts & Crafts Store, Asian Restaurant, Athletics & Sports, Australian Restaurant, Auto Garage, Auto Workshop, BBQ Joint, Bagel Shop, Bakery, Bar, Beach, Bed & Breakfast, Beer Bar, Beer Garden, Beer Store, Bike Shop, Bistro, Boarding House, Bookstore, Botanical Garden, Boutique, Bowling Alley, Boxing Gym, Brasserie, Brazilian Restaurant, Breakfast Spot, Brewery, Building, Burger Joint, Burrito Place, Bus Station, Bus Stop, Butcher, Café, Camera Store, Canal, Candy Store, Caribbean Restaurant, Caucasian Restaurant, Cemetery, Champagne Bar, Cheese Shop, Child Care Service, Chinese Restaurant, Chocolate Shop, Church, Circus School, Climbing Gym, Clothing Store, Cocktail Bar, Coffee Shop, College Quad, Comedy Club, Comfort Food Restaurant, Community Center, Concert Hall, Construction & Landscaping, Convenience Store, Convention Center, Cosmetics Shop, Costume Shop, Creperie, Cricket Ground, Cupcake Shop, Cycle Studio, Dance Studio, Deli / Bodega, Department Store, Dessert Shop, Dim Sum Restaurant, Diner, Discount Store, Distillery, Dog Run, Doner Restaurant, Donut Shop, Dry Cleaner, Electronics Store, English Restaurant, Escape Room, Ethiopian Restaurant, Event Space, Exhibit, Falafel Restaurant, Farmers Market, Fast Food Restaurant, Filipino Restaurant, Film Studio, Fish & Chips Shop, Fish Market, Fishing Store, Flea Market, Flower Shop, Food & Drink Shop, ...]
Index: []

### 5. Results and Discussion

The neighbourhoods of London are very mulitcultural. There are a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to take a step further in this direction by having a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores. The main modes of transport seem to be Buses and trains. For leisure, the neighbourhoods are set up to have lots of parks, golf courses, zoo, gyms and Historic sites.

Overall, the city of London offers a multicultural, diverse and certainly an entertaining experience.

### 6. Conclusion

The purpose of this project was to explore the city of London and see how attractive it is to potential tourists and migrants. We explored the city based on their postal codes and then extrapolated the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.

We could see that each of the neighbourhoods in the city of London has a wide variety of experiences to offer which is unique in it's own way. The cultural diversity is quite evident which also gives the feeling of a sense of inclusion.

London seems to offer a vacation stay or a romantic gateaway with a lot of places to explore, beautiful landscapes and a wide variety of culture.Overall.